### PR TITLE
Handle (annotate, refine) artificial trees like normal trees

### PR DESCRIPTION
**Pull request by d367wang**
This PR combines changes in https://github.com/opprop/checker-framework-inference/pull/249 , as well as other adaptations to fix issues introduced by non-null artificial tree paths.

It depends on CF pull request https://github.com/opprop/checker-framework/pull/162

### CODE REVIEW

File: src/checkers/inference/InferenceVisitor.java [hunk 1]
```diff
@@ -853,4 +854,33 @@ protected InferenceValidator createTypeValidator() {
     protected void checkConstructorResult(
             AnnotatedExecutableType constructorType, ExecutableElement constructorElement) {}
     
+    @Override
+    public Void visitUnary(UnaryTree node, Void p) {
+        if (!infer) {
+            return super.visitUnary(node, p);
+        }
+
+        switch (node.getKind()) {
```
Comment (d367wang) [inline src/checkers/inference/InferenceVisitor.java:863]: Add `// fall through ` to suppress error-prone warning

File: src/checkers/inference/InferenceAnnotatedTypeFactory.java [hunk 2]
```diff
@@ -576,5 +580,26 @@ protected InferenceViewpointAdapter createViewpointAdapter() {
     protected Set<? extends AnnotationMirror> getDefaultTypeDeclarationBounds() {
         return realTypeFactory.getQualifierHierarchy().getTopAnnotations();
     }
+
+    /**
```
Comment (d367wang) [inline src/checkers/inference/InferenceAnnotatedTypeFactory.java:584]: Forget to remove

File: src/checkers/inference/VariableAnnotator.java [hunk 3]
```diff
@@ -1175,12 +1175,15 @@ private void shallowAnnotateArray(AnnotatedTypeMirror type, Tree tree, int level
...
-            if (pathToTopLevelTree != null) {
-                ASTRecord astRecord = ASTPathUtil.getASTRecordForPath(inferenceTypeFactory, pathToTopLevelTree).newArrayLevel(level);
+
+            assert pathToTopLevelTree != null;
+            ASTRecord astRecord = ASTPathUtil.getASTRecordForPath(inferenceTypeFactory, pathToTopLevelTree);
+            if (astRecord != null) {
+                astRecord = astRecord.newArrayLevel(level);
                 location = new AstPathLocation(astRecord);
             } else {
+                // astRecord for `topLevelTree` is null when it's an artificial array creation tree like for
```
Comment (wmdietl) [inline src/checkers/inference/VariableAnnotator.java:1185]: Does this branch actually get triggered? It's a bit odd that the comment moved here.
Can you add a test case that triggers this and use isHackMode to guard this work-around?

Comment (d367wang) [reply src/checkers/inference/VariableAnnotator.java:1185]: The else branch is triggered when calling `ASTPathUtil.getASTRecordForPath` on an artificial tree. An example is methods with varargs.
```
void foo(Object o1, Object o2) {
        bar();
    }

    void bar(Object... args) {
        
    }
```

This seems more of a bug than an issue that needs to be hacked. And every other use of `ASTPathUtil.getASTRecordForPath` check if null is returned.

**Review by wmdietl — APPROVED** [review]
Thanks! I'll merge once CF #162 is green again.

**Pull request merged**